librad/src/meta/project.rs

```diff
@@ -17,6 +21,65 @@ pub fn default_branch() -> String {
     DEFAULT_BRANCH.into()
 }
 
+#[derive(Debug, Fail)]
+pub enum Error {
+    #[fail(display = "Cannot serialize project metadata")]
+    SerializationFailed(serde_json::error::Error),
+    #[fail(display = "Invalid UTF8")]
+    InvalidUtf8,
+    #[fail(display = "Signature already present")]
+    SignatureAlreadyPresent,
+    #[fail(display = "Signature from non maintainer")]
+    SignatureFromNonMaintainer,
+    #[fail(display = "Signature missing")]
+    SignatureMissing,
+    #[fail(display = "Signature decoding failed")]
+    SignatureDecodingFailed,
+    #[fail(display = "Signature verification failed")]
+    SignatureVerificationFailed,
+}
+
+#[derive(Debug, Fail)]
+pub enum UpdateVerificationError {
+    #[fail(display = "Non monotonic revision")]
+    NonMonotonicRevision,
+    #[fail(display = "Update without previous quorum")]
+    NoPreviousQuorum,
+    #[fail(display = "Update without current quorum")]
+    NoCurrentQuorum,
+}
+
+#[derive(Debug, Fail)]
+pub enum HistoryVerificationError {
+    #[fail(display = "Empty history")]
+    EmptyHistory,
+    #[fail(display = "Project error")]
+    ProjectError { revision: u64, error: Error },
+    #[fail(display = "Update error")]
+    UpdateError {
+        revision: u64,
+        error: UpdateVerificationError,
+    },
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct ProjectSignature {
+    pub key: PeerId,
+    pub sig: String,
+}
+
+impl ProjectSignature {
+    pub fn verify_data(&self, data: &[u8]) -> Result<(), Error> {
+        let sig =
+            Signature::from_hex_string(&self.sig).map_err(|_| Error::SignatureDecodingFailed)?;
```

Comment:
`from_hex_string` is from a different module (`device` instead of `project`) so I did not want to use `project::Error` there.
IMHO for this reason some kind of error conversion is inevitable.

Reply:
Ah, I didn't realise that it's cross-module. Git reviews can make noticing that hard sometimes :sweat_smile: